cpp/ql/src/Likely Bugs/Underspecified Functions/ImplicitFunctionDeclaration.ql

```diff
@@ -0,0 +1,158 @@
+/**
+ * @name Implicit function declaration
+ * @description An implicitly declared function is assumed to take no
+ * arguments and return an integer. If this assumption does not hold, it
+ * may lead to unpredictable behavior.
+ * @kind problem
+ * @problem.severity warning
+ * @precision medium
+ * @id cpp/implicit-function-declaration
+ * @tags correctness
+ *       maintainability
+ */
+
+import cpp
+
+pragma[inline]
+predicate arithTypesMatch(Type arg, Type parm) {
+  arg = parm
+  or
+  arg.getSize() = parm.getSize() and
+  (
+    arg instanceof IntegralOrEnumType and
+    parm instanceof IntegralOrEnumType
+    or
+    arg instanceof FloatingPointType and
+    parm instanceof FloatingPointType
+  )
+}
+
+pragma[inline]
+predicate nestedPointerArgTypeMayBeUsed(Type arg, Type parm) {
+  // arithmetic types
+  arithTypesMatch(arg, parm)
+  or
+  // conversion to/from pointers to void is allowed
+  arg instanceof VoidType
+  or
+  parm instanceof VoidType
+}
+
+pragma[inline]
+predicate pointerArgTypeMayBeUsed(Type arg, Type parm) {
+  nestedPointerArgTypeMayBeUsed(arg, parm)
+  or
+  // nested pointers
+  nestedPointerArgTypeMayBeUsed(arg.(PointerType).getBaseType().getUnspecifiedType(),
+    parm.(PointerType).getBaseType().getUnspecifiedType())
+  or
+  nestedPointerArgTypeMayBeUsed(arg.(ArrayType).getBaseType().getUnspecifiedType(),
+    parm.(PointerType).getBaseType().getUnspecifiedType())
+}
+
+pragma[inline]
```

Comment:
Every use of `pragma[inline]` leads to more relations being joined in one predicate, which makes it more likely that the join orderer will make a mistake. Removing `pragma[inline]`, on the other hand, could lead to wasted work if the optimiser isn't able to apply magic. But as long as that wasted work is at most linear in the size of the program, we don't usually consider it to be a problem.

Reply:
That makes sense. The `pragma[inline]`s was a fix to a performance problem I had earlier during the development of the query, which actually ended up being fixed by a different set of changes. I will factor out the bodies of `cpp/futile-params`, `cpp/too-few-arguments` and `cpp/mistyped-function-arguments` and remove all of these predicates (and their `pragma[inline]` annotations) later today!

Reply:
After your latest refactoring I can see that all the `pragma[inline]` came from existing queries. I see no need to change those as part of this PR -- that can be done separately if we see a need for it.